### PR TITLE
[feat] Separate review issues

### PR DESCRIPTION
**Pull request by Adamih**
#789

### CODE REVIEW

**Comment by slarse** [comment]
Thanks @Adamih @ZinoKader, at first glance it looks like this will do the job! Nice attention to detail with the documentation as well :)

Just closing and opening again to trigger the GitLab system tests here...

**Comment by slarse** [comment]
Looks like a GitLab system test needs to be updated, it expects a single issue with a static name.

**Review by slarse — CHANGES_REQUESTED** [review]
One of the GitLab system tests is failing as the assertion is out-of-date. It's assuming a single issue to be opened with a static name, which clearly doesn't hold after this improvement. So you need to go in there and tweak the test a bit!

**Comment by ZinoKader** [comment]
I am unfortunately unable to run the GitLab system tests and it may have to do with my M1 macbook... [This is the issue](https://forum.gitlab.com/t/gitlab-ce-docker-on-apple-m1-stuck-on-recipe-gitlab-database-migrations/49297/3) I am experiencing.

**Comment by Adamih** [comment]
This is the best solution I could come up with

```python
assert_num_issues(STUDENT_TEAMS, [assignment_name], 1)
# For each reviewed student
for reviewed_student in STUDENT_TEAMS:
    # Get all potential reviewing students
    reviewing_students = (
         s for s in STUDENT_TEAMS if s != reviewed_student
    )
    # Match title
    assert_issue_title_match(
        [reviewing_student],
        [assignment_name],
        "^Review issue \((?:" + "|".join(reviewing_students) + ")\)$",    # Any of potential reviewing students
    )
```

````python
def assert_issue_title_match(student_teams, assignment_names, pattern):
    def assertion(project):
        issues = project.issues.list(all=True)
        for actual_issue in issues:
            if re.match(pattern, actual_issue.title):
                return
        assert False, "no issue title matching the specified pattern"

    _assert_on_projects(student_teams, assignment_names, assertion)
```

This is the best idea I have but it still seems a bit finicky.

File: system_tests/gitlab/_helpers/asserts.py [hunk 1]
```diff
@@ -158,6 +159,17 @@ def assertion(project):
     _assert_on_projects(student_teams, assignment_names, assertion)
 
 
+def assert_issue_title_match(student_teams, assignment_names, pattern):
+    def assertion(project):
+        issues = project.issues.list(all=True)
+        for actual_issue in issues:
+            if re.match(pattern, actual_issue.title):
+                return
+        assert False, "no issue title matching the specified pattern"
```
[the remark this answers is not in the document]
Comment (slarse) [reply system_tests/gitlab/_helpers/asserts.py:168]: Curious decision to put all assertions (which are integral to the tests) in a module separate from the tests. Yesterday me is full of surprises.

I'm not saying you should do anything about that, I'll open a separate issue about it.

File: system_tests/gitlab/test_gitlab_system.py [hunk 2]
```diff
@@ -618,12 +619,15 @@ def test_assign_one_review(self, with_student_repos, tmpdir):
...
-        )
+        for reviewed_team in STUDENT_TEAMS:
+            any_other_team_pattern = "|".join(
+                s.name for s in STUDENT_TEAMS if s != reviewed_team
+            )
+            assert_issue_title_match(
+                [reviewed_team],
+                [assignment_name],
+                rf"^Peer review \((?:{any_other_team_pattern})\)$",
+            )
```
Comment (slarse) [inline system_tests/gitlab/test_gitlab_system.py:630]: I would suggest to weaken the assertions here for the sake of simplicity. System tests are more about the bigger picture than minute details, so I think it's perfectly fine to have looser assertions than one normally would in a lower level test.

```suggestion
    assert_issue_title_match(STUDENT_TEAMS, [assignment_name], r"^Peer review.*$")
```

**Review by slarse — CHANGES_REQUESTED** [review]
I've been preoccupied over the past few days, hence the communications blackout.

Anyway, this looks very good, I'd rate this a very high quality PR. I have some comments on simplifications that I'd like to see and some miscellaneous stuff that you couldn't possibly have known. Overall I'm very happy with this contribution and I think we can merge once you've addressed my comments.

One more thing: this branch is out of date with the base branch, could you update it (i.e. merge master into it)?

File: system_tests/gitlab/_helpers/asserts.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (slarse) [reply system_tests/gitlab/_helpers/asserts.py:168]: #1020

**Comment by slarse** [comment]
Closing as stale, feel free to revisit.

**Pull request closed without merging**